beacon-chain/cache/sync_subnet_ids.go

```diff
@@ -0,0 +1,120 @@
+	"github.com/prysmaticlabs/prysm/shared/sliceutil"
+)
+
+type syncSubnetIDs struct {
+	sCommittee    *cache.Cache
+	sCommiteeLock sync.RWMutex
+}
+
+// SyncSubnetIDs for sync committee participant.
+var SyncSubnetIDs = newSyncSubnetIDs()
+
+func newSyncSubnetIDs() *syncSubnetIDs {
+	epochDuration := time.Duration(params.BeaconConfig().SlotsPerEpoch.Mul(params.BeaconConfig().SecondsPerSlot))
+	subLength := epochDuration * time.Duration(params.BeaconConfig().EpochsPerSyncCommitteePeriod)
+	persistentCache := cache.New(subLength*time.Second, epochDuration*time.Second)
+	return &syncSubnetIDs{sCommittee: persistentCache}
+}
+
+// GetSyncCommitteeSubnets retrieves the sync committee subnet and expiration time of that validator's
+// subscription.
+func (s *syncSubnetIDs) GetSyncCommitteeSubnets(pubkey []byte, epoch types.Epoch) ([]uint64, types.Epoch, bool, time.Time) {
+	s.sCommiteeLock.RLock()
+	defer s.sCommiteeLock.RUnlock()
+
+	id, duration, ok := s.sCommittee.GetWithExpiration(keyBuilder(pubkey, epoch))
+	if !ok {
+		return []uint64{}, 0, ok, time.Time{}
+	}
+	// Retrieve the slot from the cache.
+	idxs, ok := id.([]uint64)
+	if !ok {
+		return []uint64{}, 0, ok, time.Time{}
+	}
+	// If no committees are saved, we return
+	// nothing.
+	if len(idxs) <= 1 {
+		return []uint64{}, 0, ok, time.Time{}
+	}
+	return idxs[1:], types.Epoch(idxs[0]), ok, duration
+}
+
+// GetAllSubnets retrieves all the non-expired subscribed subnets of all the validators
+// in the cache. This method also takes the epoch as an argument to only retrieve
+// assingments for epochs that have happened.
+func (s *syncSubnetIDs) GetAllSubnets(currEpoch types.Epoch) []uint64 {
+	s.sCommiteeLock.RLock()
+	defer s.sCommiteeLock.RUnlock()
+
+	itemsMap := s.sCommittee.Items()
+	var committees []uint64
+
+	for _, v := range itemsMap {
+		if v.Expired() {
+			continue
+		}
+		idxs, ok := v.Object.([]uint64)
+		if !ok {
+			continue
+		}
+		if len(idxs) <= 1 {
+			continue
+		}
+		// Check if the subnet is valid in the current epoch.
+		if types.Epoch(idxs[0]) > currEpoch {
+			continue
+		}
+		// Ignore the first index as that represents the
+		// epoch of the validator's assignments.
+		committees = append(committees, idxs[1:]...)
+	}
+	return sliceutil.SetUint64(committees)
+}
+
+// AddSyncCommitteeSubnets adds the relevant committee for that particular validator along with its
+// expiration period. An Epoch argument here denotes the epoch from which the sync committee subnets
+// will be active.
+func (s *syncSubnetIDs) AddSyncCommitteeSubnets(pubkey []byte, epoch types.Epoch, comIndex []uint64, duration time.Duration) {
+	s.sCommiteeLock.Lock()
+	defer s.sCommiteeLock.Unlock()
+	subComCount := params.BeaconConfig().SyncCommitteeSubnetCount
+	// To join a sync committee subnet, select a random number of epochs before the end of the
+	// current sync committee period between 1 and SYNC_COMMITTEE_SUBNET_COUNT, inclusive.
+	diff := (rand.NewGenerator().Uint64() % subComCount) + 1
```

Comment:
Is this a spec thing? This seems very non-deterministic

Reply:
@nisdas can you answer this? I can't find the exact line in the p2p spec where this is specified

Reply:
Spec
> To join a sync committee subnet, select a random number of epochs before the end of the current sync committee period between 1 and SYNC_COMMITTEE_SUBNET_COUNT, inclusive. Validators should join their member subnet at the beginning of the epoch they have randomly selected. For example, if the next sync committee period starts at epoch 853,248 and the validator randomly selects an offset of 3, they should join the subnet at the beginning of epoch 853,245. Validators should leverage the lookahead period on sync committee assignments so that they can join the appropriate subnets ahead of their assigned sync committee period.

https://github.com/ethereum/eth2.0-specs/blob/master/specs/altair/validator.md#sync-committee-subnet-stability